Assets/MixedRealityToolkit.Tests/PlayModeTests/RenderingTests.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#if !WINDOWS_UWP
+// When the .NET scripting backend is enabled and C# projects are built
+// The assembly that this file is part of is still built for the player,
+// even though the assembly itself is marked as a test assembly (this is not
+// expected because test assemblies should not be included in player builds).
+// Because the .NET backend is deprecated in 2018 and removed in 2019 and this
+// issue will likely persist for 2018, this issue is worked around by wrapping all
+// play mode tests in this check.
+
+using Microsoft.MixedReality.Toolkit.Rendering;
+using Microsoft.MixedReality.Toolkit.Utilities;
+using NUnit.Framework;
+using System.Collections;
+using UnityEngine;
+using UnityEngine.TestTools;
+
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    public class RenderingTests : BasePlayModeTests
+    {
+        private const float VisualizatonWaitTime = 1.0f;
+        private readonly Vector3 LightDirection = new Vector3(-0.5f, -0.5f, 0.0f);
+
+        /// <summary>
+        /// Tests if <see cref="MaterialInstance"/> can be instantiated and manipulated at runtime.
+        /// </summary>
+        [UnityTest]
+        public IEnumerator TestMaterialInstance()
+        {
+            // Light only used for developers to visually inspect test.
+            var light = InstantiateDirectionalLight(LightDirection);
+
+            var gameObject = GameObject.CreatePrimitive(PrimitiveType.Cube);
+            var redMaterial = InstantiateStandardShaderMaterial(Color.red);
+            gameObject.GetComponent<Renderer>().sharedMaterial = redMaterial;
+
+            // Test adding a MaterialInstance.
+            var materialInstance = gameObject.AddComponent<MaterialInstance>();
+
+            // Test altering properties.
+            materialInstance.Material.color = Color.green;
+            materialInstance.Material.mainTexture = Texture2D.blackTexture;
+
+            // Test multiple materials and new material tracking.
+            var blueMaterial = InstantiateStandardShaderMaterial(Color.blue);
+            gameObject.GetComponent<Renderer>().sharedMaterials = new Material[] { redMaterial, blueMaterial };
+
+            yield return null;
+
+            Assert.AreEqual(2, materialInstance.Materials.Length, "The MaterialInstance should contain two materials.");
```

Comment:
Since this deals with lifetime management, it would give more confidence if we stressed these tests a bit more, something like creating 100 objects, changing the materials every frame on each object, then destroying all 100 and ensuring nothing is leaked.

Reply:
That's a great idea! Let me see if there is an API to query the scene material count. If so, I can add this in a future PR. :)